src/metabase/query_processor/util/nest_query.clj

```diff
@@ -1,352 +1,120 @@
-            (assoc :source-query source-query))))))
-
-(mu/defn nest-expressions :- ::mbql.s/SourceQuery
-  "Pushes the `:source-table`/`:source-query`, `:expressions`, and `:joins` in the top-level of the query into a
-  `:source-query` and updates `:expression` references and `:field` clauses with `:join-alias`es accordingly. See
-  tests for examples. This is used by the SQL QP to make sure expressions happen in a subselect."
-  [inner-query :- ::mbql.s/SourceQuery]
-  (-> inner-query
-      nest-expressions*
-      add/add-alias-info))
+   [metabase.util.performance :refer [select-keys some]]))
+
+;; Mark all Fields at the new top level as `:qp/ignore-coercion` so QP implementations know not to apply coercion
+;; or whatever to them a second time.
+;; In fact, we don't mark all Fields, only the ones we deem coercible. Marking all would make a bunch of tests
+;; fail, but it might still make sense. For example, #48721 would have been avoided by unconditional marking.
+
+(defn- contains-expression-refs? [location]
+  (lib.util.match/match-one location :expression))
+
+(defn- should-nest-expressions? [query path]
+  (and (lib.walk/apply-f-for-stage-at-path lib/mbql-stage? query path)
+       (some (fn [f]
+               (contains-expression-refs? (lib.walk/apply-f-for-stage-at-path f query path)))
+             [lib/breakouts
+              lib/aggregations
+              lib/order-bys])))
+
+(def ^:private first-stage-keys
+  #{:expressions :joins :source-table :source-card})
+
+;; TODO (Cam 10/22/25) -- somewhat duplicated with/copied
+;; from [[metabase.query-processor.util.transformations.nest-breakouts/fields-used-in-breakouts-aggregations-or-expressions]]
+(defn- fields-needed-by-second-stage [stage]
+  (let [stage' (apply dissoc stage first-stage-keys)
+        refs (volatile! (transient []))]
+    ;; temporarily disable enforcement since this stage is technically invalid since we removed all
+    ;; the [[first-stage-keys]]
+    (binding [lib.schema/*HACK-disable-ref-validation* true]
+      (lib.walk/walk-clauses-in-stage
+       stage'
+       (fn [clause]
+         (u/prog1 clause
+           (when (lib/clause-of-type? clause #{:field :expression})
+             (vswap! refs conj! clause))))))
+    ;; use an ordered set so we preserve the order we saw things when we walked the query so the fields we return are
+    ;; determinate. Otherwise tests using this are liable to be flaky because results can change because test metadata
+    ;; has randomly generated IDs
+    (into
+     []
+     (m/distinct-by (fn [[tag opts id-or-name]]
+                      [tag
+                       (select-keys opts [:join-alias :temporal-unit :bucketing])
+                       id-or-name]))
+     (persistent! @refs))))
+
+(defn- new-first-stage [stage]
+  (-> stage
+      (select-keys (conj first-stage-keys :lib/type))
+      (assoc :fields (lib/fresh-uuids (vec (fields-needed-by-second-stage stage))))))
```

Comment:
`fields-needed-by-second-stage` always returns a vector. Did you mean to put the `vec` outside `lib/fresh-uuids`?